### DISCUSSION TITLE
Representing a dictionary with fixed and dynamic keys at the root level

### DISCUSSION THREAD

**Discussion by TimekillerTK** [pydantic/pydantic#8739] [Question]
How can I use Pydantic to represent the following dictionary, assuming there can be an arbitrary number of objects of type C (in this example also D & E), each with an arbitrary name at the root level?

```python
example_dict = {
    "A": ["one", "two", "three"],
    "B": {"key_b": "value"},
    "C": {
        "name": "value1",
        "tags": ["tag_name"],
        "data": {"key_data": "value1"},
    },
    "D": {
        "name": "value2",
        "tags": ["tag_name"],
        "data": {"key_data": "value2"},
    },
    "E": {
        "name": "value3",
        "tags": ["tag_name"],
        "data": {"key_data": "value3"},
    },
}
```

Normally, the following approach can be used, which works if there's a known number of objects at the root level of the dictionary (in this case 3). Note that C, D and E are all of type Common.

_This approach fails to represent the dictionary, if there's more objects of type Common (imagine if there's F,G and H, etc)_

```python
class KeyData(BaseModel):
    key_data: str


class ClassB(BaseModel):
    key_b: str


class Common(BaseModel):
    name: str
    tags: list[str]
    data: KeyData


class ExampleDict(BaseModel):
    A: list[str]
    B: ClassB
    C: Common
    D: Common
    E: Common


load_example = ExampleDict(**example_dict)
print(load_example.model_dump())
# {'A': ['one', 'two', 'three'], 'B': {'key_b': 'value'}, 'C': {'name': 'value1', 'tags': ['tag_name'], 'data': {'key_data': 'value1'}}, 'D': {'name': 'value2', 'tags': ['tag_name'], 'data': {'key_data': 'value2'}}, 'E': {'name': 'value3', 'tags': ['tag_name'], 'data': {'key_data': 'value3'}}}
```

Or alternatively, a `RootModel` can be used instead of a `BaseModel` to allow for an arbitrary amount of objects with arbitrary names. However this will not work, because the `RootModel` does not allow any fields other than `root`.

```python
class KeyData(BaseModel):
    key_data: str


class ClassB(BaseModel):
    key_b: str


class Common(BaseModel):
    name: str
    tags: list[str]
    data: KeyData


class ExampleDict(RootModel):
    A: list[str]
    B: ClassB
    root: dict[str, Common]


load_example = ExampleDict(**example_dict)
print(load_example.model_dump())
# NameError: Unexpected field with name 'A'; only 'root' is allowed as a field of a `RootModel`
```

What approach can be used in this situation, to represent a dictionary with both fixed and arbitrary keys at the root level?

**Comment by rafalkrupinski** [comment]
Pydantic doesn't offer dynamic fields at the root level of you model class, but it does support dynamic properties in your data.
You can declare your "C" type in `__model_extra__` special field and the values appear in `model_extra` synthetic field.
